toolchain/check/testdata/function/generic/deduce.carbon

```diff
@@ -0,0 +1,630 @@
+
+// --- deduce_explicit.carbon
+
+library "fail_todo_deduce_explicit";
+
+fn ExplicitGenericParam(T:! type) -> T*;
+
+fn CallExplicitGenericParam() -> i32* {
+  return ExplicitGenericParam(i32);
+}
+
+// --- fail_todo_explicit_vs_deduced.carbon
+
+library "fail_todo_explicit_vs_deduced";
+
+class A {}
+
+fn ExplicitAndAlsoDeduced(T:! type, x: T) -> T*;
+
+// TODO: This should presumably be accepted. We shouldn't deduce values for parameters with explicitly-specified values.
+fn CallExplicitAndAlsoDeduced(n: i32) -> i32* {
+  // CHECK:STDERR: fail_todo_explicit_vs_deduced.carbon:[[@LINE+7]]:10: ERROR: Inconsistent deductions for value of generic parameter `T`.
+  // CHECK:STDERR:   return ExplicitAndAlsoDeduced(A, {});
+  // CHECK:STDERR:          ^~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR: fail_todo_explicit_vs_deduced.carbon:[[@LINE-7]]:1: While deducing parameters of generic declared here.
+  // CHECK:STDERR: fn ExplicitAndAlsoDeduced(T:! type, x: T) -> T*;
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  return ExplicitAndAlsoDeduced(A, {});
+}
+
+// --- fail_todo_deduce_implicit.carbon
+
+library "fail_todo_deduce_implicit";
+
+fn ImplicitGenericParam[T:! type](x: T) -> T*;
+
+fn CallImplicitGenericParam(n: i32) -> i32* {
+  // CHECK:STDERR: fail_todo_deduce_implicit.carbon:[[@LINE+4]]:10: ERROR: Semantics TODO: `Call with implicit parameters`.
+  // CHECK:STDERR:   return ImplicitGenericParam(n);
+  // CHECK:STDERR:          ^~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  return ImplicitGenericParam(n);
+}
+
+// --- fail_todo_deduce_nested.carbon
+
+library "fail_todo_deduce_nested";
+
+fn TupleParam[T:! type](x: (T, i32));
+
+fn CallTupleParam() {
+  // CHECK:STDERR: fail_todo_deduce_nested.carbon:[[@LINE+7]]:3: ERROR: Cannot deduce value for generic parameter `T`.
+  // CHECK:STDERR:   TupleParam((1, 2));
+  // CHECK:STDERR:   ^~~~~~~~~~~
+  // CHECK:STDERR: fail_todo_deduce_nested.carbon:[[@LINE-6]]:1: While deducing parameters of generic declared here.
+  // CHECK:STDERR: fn TupleParam[T:! type](x: (T, i32));
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  TupleParam((1, 2));
+}
+
+fn StructParam[T:! type](x: {.a: T, .b: i32});
+
+fn CallStructParam() {
+  // CHECK:STDERR: fail_todo_deduce_nested.carbon:[[@LINE+7]]:3: ERROR: Cannot deduce value for generic parameter `T`.
+  // CHECK:STDERR:   StructParam({.a = 1, .b = 2});
+  // CHECK:STDERR:   ^~~~~~~~~~~~
+  // CHECK:STDERR: fail_todo_deduce_nested.carbon:[[@LINE-6]]:1: While deducing parameters of generic declared here.
+  // CHECK:STDERR: fn StructParam[T:! type](x: {.a: T, .b: i32});
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  StructParam({.a = 1, .b = 2});
+}
+
+// --- fail_deduce_incomplete.carbon
+
+library "fail_deduce_incomplete";
+
+fn ImplicitNotDeducible[T:! type, U:! type](x: T) -> U;
```

Comment:
This is probably future work, but it would really be nice if we could diagnose that `U` isn't deducible for any possible usage, without having to see a usage.

Reply:
Added a TODO.